### PR TITLE
Add pagination to events panel

### LINKED ISSUE #7: Upcoming Events / Previous Events Pagination
Don't want to show a million things at one time

### PR DESCRIPTION
**Pull request by zakpatterson**
Fixes #7 

Max events per page is set to 5, if there are more than 5 events, a search bar is displayed which filters events. 

If there are more than 5 events, pagination buttons are available at the bottom.

Filtering is applied before pagination.

### CODE REVIEW

**Comment by zakpatterson** [comment]
Assuming you use squash merging by default

**Comment by samwightt** [comment]
@zakpatterson Oops I use just the 'merge pull request' button, but idk if it matters too much, just that the stuff gets merged in with a commit that links back to this if we need it.

Feel free to deploy when you want to btw! We don't have any standard code review process AFAIK.

**Comment by zakpatterson** [comment]
I'm happy to have someone else just take a quick look at it. I also don't have merge rights.  
I'd suggest only allowing squash merges, to prevent commits of brokenness from getting into the main branch.

2022-09-12-151127_869x437_scrot

File: packages/site/src/lib/components/Agenda.svelte [hunk 1]
```diff
@@ -11,6 +11,16 @@
...
+  
+  const maxPerPage: number = 5;
+
+  let pageNumber: number = 1;
+  let filteredEvents: Event[] = events;
+  let numPages: number = Math.ceil(filteredEvents.length / maxPerPage);
+
+  let currentPageEvents: Event[];
+
+  let searchValue: string = '';
```
Comment (samwightt) [inline packages/site/src/lib/components/Agenda.svelte]:
```suggestion
  let pageNumber: number = 1;
  let searchValue: string = '';
  
  const eventMatches = (ev: Event, search: string): boolean => ev.name.toLowerCase().includes(search.toLowerCase())
  const filterEvents = (evs: Event[], search: string): Event[] => search.trim() === '' ? evs : evs.filter(ev => eventMatches(ev, search))
  
  $: filteredEvents: Event[] = filterEvents(events, searchString)
  
  const paginateEvents = (evs: Event[], pageNum: number, maxPerPage: number): Event[] => {
    const start = (pageNum - 1) * maxPerPage
    const end = start + maxPerPage
    return evs.slice(start, end)
  }
  
  $: currentPageEvents: Event[] = paginateEvents(filteredEvents, pageNumber, maxPerPage)
  
  $: numPages: number = Math.ceil(filteredEvents.length / maxPerPage);
```

This should use the reactive declaration syntax. Saves you a toooon of work. You have to explicitly reference the variables in the statement or block. So the reason why I pass all the variables as arguments to `paginateEvents` and `filterEvents` is to make sure each of the `$` blocks knows its dependencies.

File: packages/site/src/lib/components/Agenda.svelte [hunk 2]
```diff
@@ -20,19 +30,60 @@
...
+    pageNumber += 1;
+    updatePagedEvents();
+  };
+
+  const updateEvents = (): void => {
+    filteredEvents = getFilteredEvents()
+    updatePagedEvents();
+  };
+
+  updatePagedEvents();
```
Comment (samwightt) [inline packages/site/src/lib/components/Agenda.svelte]:
```suggestion
  $: canAdvance: number = pageNumber < numPages
  $: canGoBack: number = pageNumber > 1

  const goBack = (): void => {
    if (canGoBack) {
      pageNumber -= 1;
    }
  };

  const advance = (): void => {
    if (canAdvance) {
      pageNumber += 1;
    }
  };
```

Probably should check this here, plus this gives these concepts a name.

File: packages/site/src/lib/components/Agenda.svelte [hunk 3]
```diff
@@ -46,6 +97,19 @@
         </a>
       {/each}
     </div>
+    {#if numPages > 1}
+      <div class="right group">
+        {#if pageNumber < numPages}
```
Comment (samwightt) [inline packages/site/src/lib/components/Agenda.svelte]:
```suggestion
        {#if canAdvance}
```

File: packages/site/src/lib/components/Agenda.svelte [hunk 4]
```diff
@@ -46,6 +97,19 @@
...
+          <div><button type="button" on:click={advance}>Next</button></div>
+        {/if}
+        <div>Page {pageNumber} of {numPages}</div>
+        {#if pageNumber > 1}
```
Comment (samwightt) [inline packages/site/src/lib/components/Agenda.svelte]:
```suggestion
        {#if canGoBack}
```

**Review by samwightt — CHANGES_REQUESTED** [review]
Just notes on using the reactive declaration syntax. Once you use it you seriously won't want to go back 😅

**Pull request merged**